apps/cdn/src/cache/key.rs

```diff
@@ -0,0 +1,175 @@
+	}
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ImageFile {
+	pub name: ImageFileName,
+	pub extension: ImageFileExtension,
+	pub is_static: bool,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ImageFileName {
+	One,
+	Two,
+	Three,
+	Four,
+}
+
+impl Display for ImageFileName {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			ImageFileName::One => write!(f, "1x"),
+			ImageFileName::Two => write!(f, "2x"),
+			ImageFileName::Three => write!(f, "3x"),
+			ImageFileName::Four => write!(f, "4x"),
+		}
+	}
+}
+
+impl FromStr for ImageFileName {
+	type Err = anyhow::Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		match s {
+			"1x" => Ok(Self::One),
+			"2x" => Ok(Self::Two),
+			"3x" => Ok(Self::Three),
+			"4x" => Ok(Self::Four),
+			_ => anyhow::bail!("invalid image file name"),
+		}
+	}
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ImageFileExtension {
+	Avif,
+	Gif,
+	Png,
+	Webp,
+}
+
+impl Display for ImageFileExtension {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			ImageFileExtension::Avif => write!(f, "avif"),
+			ImageFileExtension::Gif => write!(f, "gif"),
+			ImageFileExtension::Png => write!(f, "png"),
+			ImageFileExtension::Webp => write!(f, "webp"),
+		}
+	}
+}
+
+impl FromStr for ImageFileExtension {
+	type Err = anyhow::Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		match s {
+			"avif" => Ok(Self::Avif),
+			"gif" => Ok(Self::Gif),
+			"png" => Ok(Self::Png),
+			"webp" => Ok(Self::Webp),
+			_ => anyhow::bail!("invalid file extension"),
+		}
+	}
+}
+
+impl FromStr for ImageFile {
+	type Err = anyhow::Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		if let Some((mut name, ext)) = s.split_once('.') {
+			let extension = ImageFileExtension::from_str(ext)?;
+
+			let is_static = if let Some(n) = name.strip_suffix("_static") {
+				name = n;
+				true
+			} else {
+				false
+			};
+
+			Ok(Self {
+				name: ImageFileName::from_str(name)?,
+				extension,
+				is_static,
+			})
+		} else {
+			Ok(Self {
+				name: ImageFileName::from_str(s)?,
+				extension: ImageFileExtension::Webp,
+				is_static: false,
+			})
+		}
+	}
+}
+
+impl Display for ImageFile {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		if self.is_static {
+			write!(f, "{}_static.{}", self.name, self.extension)
+		} else {
+			write!(f, "{}.{}", self.name, self.extension)
+		}
+	}
+}
+
+impl serde::Serialize for ImageFile {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: serde::Serializer,
+	{
+		serializer.serialize_str(&self.to_string())
+	}
+}
+
+impl<'de> serde::Deserialize<'de> for ImageFile {
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		let s = String::deserialize(deserializer)?;
+		Ok(s.parse().map_err(|e| serde::de::Error::custom(e))?)
+	}
+}
```

Comment:
why do we need serde?

Reply:
axum uses serde to parse path parameters